ambari-server/src/main/java/org/apache/ambari/server/controller/internal/UpgradePlanResourceProvider.java

```diff
@@ -0,0 +1,342 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.ambari.server.controller.internal;
+
+import static org.apache.ambari.server.controller.utilities.PropertyHelper.EXTERNAL_PATH_SEP;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.EnumSet;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import org.apache.ambari.server.AmbariException;
+import org.apache.ambari.server.StaticallyInject;
+import org.apache.ambari.server.controller.AmbariManagementController;
+import org.apache.ambari.server.controller.spi.NoSuchParentResourceException;
+import org.apache.ambari.server.controller.spi.NoSuchResourceException;
+import org.apache.ambari.server.controller.spi.Predicate;
+import org.apache.ambari.server.controller.spi.Request;
+import org.apache.ambari.server.controller.spi.RequestStatus;
+import org.apache.ambari.server.controller.spi.Resource;
+import org.apache.ambari.server.controller.spi.ResourceAlreadyExistsException;
+import org.apache.ambari.server.controller.spi.SystemException;
+import org.apache.ambari.server.controller.spi.UnsupportedPropertyException;
+import org.apache.ambari.server.orm.dao.UpgradePlanDAO;
+import org.apache.ambari.server.orm.entities.UpgradePlanDetailEntity;
+import org.apache.ambari.server.orm.entities.UpgradePlanEntity;
+import org.apache.ambari.server.security.authorization.RoleAuthorization;
+import org.apache.ambari.server.state.Cluster;
+import org.apache.ambari.server.state.stack.upgrade.Direction;
+import org.apache.ambari.server.state.stack.upgrade.UpgradeType;
+import org.apache.commons.collections.CollectionUtils;
+import org.apache.commons.lang.BooleanUtils;
+import org.apache.commons.lang.StringUtils;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Sets;
+import com.google.gson.Gson;
+import com.google.gson.annotations.SerializedName;
+import com.google.gson.reflect.TypeToken;
+import com.google.inject.Inject;
+
+/**
+ * Manages upgrade plans.
+ */
+@StaticallyInject
+public class UpgradePlanResourceProvider extends AbstractControllerResourceProvider {
+
+  private static final String UPGRADE_PLAN = "UpgradePlan";
+
+  protected static final String UPGRADE_PLAN_ID = UPGRADE_PLAN + EXTERNAL_PATH_SEP + "id";
+  protected static final String UPGRADE_PLAN_CLUSTER_NAME = UPGRADE_PLAN + EXTERNAL_PATH_SEP + "cluster_name";
+
+  public static final String UPGRADE_PLAN_TYPE = UPGRADE_PLAN + EXTERNAL_PATH_SEP + "upgrade_type";
```

Comment:
why not define `UPGRADE_PLAN + EXTERNAL_PATH_SEP` as another static constant, and use it instead of concatenating each time 3 strings?

Reply:
Will do.  I typically do just straight strings but this was a clone of some other resource provider and is supposed to help the ApiDoc stuff (but I don't know how)